core/runtime.rs

```diff
@@ -75,31 +85,134 @@ struct IsolateAllocations {
+    self.0.deref_mut()
+  }
+}
+
+/// This struct contains the [`JsRuntimeState`] and [`v8::OwnedIsolate`] that are required
+/// to do an orderly shutdown of V8. We keep these in a separate struct to allow us to control
+/// the destruction more closely, as snapshots require the isolate to be destroyed by the
+/// snapshot process, not the destructor.
+struct InnerIsolateState {
+  snapshotting: bool,
+  state: ManuallyDropRc<RefCell<JsRuntimeState>>,
+  v8_isolate: ManuallyDrop<v8::OwnedIsolate>,
+}
+
+impl InnerIsolateState {
+  /// Clean out the opstate and take the inspector to prevent the inspector from getting destroyed
+  /// after we're torn down the contexts. If the inspect is not correctly torn down, random crashes
+  /// happen in tests (and possibly for users using the inspector).
+  pub fn prepare_for_cleanup(&mut self) {
+    let mut state = self.state.borrow_mut();
+    let inspector = state.inspector.take();
+    state.op_state.borrow_mut().clear();
+    if let Some(inspector) = inspector {
+      assert_eq!(
+        Rc::strong_count(&inspector),
+        1,
+        "The inspector must be dropped before the runtime"
+      );
+    }
+  }
+
+  pub fn cleanup(&mut self) {
+    self.prepare_for_cleanup();
+
+    let state_ptr = self.v8_isolate.get_data(STATE_DATA_OFFSET);
+    // SAFETY: We are sure that it's a valid pointer for whole lifetime of
+    // the runtime.
+    _ = unsafe { Rc::from_raw(state_ptr as *const RefCell<JsRuntimeState>) };
+
+    let module_map_ptr = self.v8_isolate.get_data(MODULE_MAP_DATA_OFFSET);
+    // SAFETY: We are sure that it's a valid pointer for whole lifetime of
+    // the runtime.
+    _ = unsafe { Rc::from_raw(module_map_ptr as *const RefCell<ModuleMap>) };
+
+    self.state.borrow_mut().destroy_all_realms();
+
+    debug_assert_eq!(Rc::strong_count(&self.state), 1);
+  }
+
+  pub fn prepare_for_snapshot(mut self) -> v8::OwnedIsolate {
+    self.cleanup();
+    // SAFETY: We're copying out of self and then immediately forgetting self
+    let (state, isolate) = unsafe {
+      (
+        ManuallyDrop::take(&mut self.state.0),
+        ManuallyDrop::take(&mut self.v8_isolate),
+      )
+    };
+    std::mem::forget(self);
+    drop(state);
+    isolate
+  }
+}
+
+impl Drop for InnerIsolateState {
+  fn drop(&mut self) {
+    self.cleanup();
+    // SAFETY: We gotta drop these
+    unsafe {
+      ManuallyDrop::drop(&mut self.state.0);
+      if self.snapshotting {
+        // Create the snapshot and just drop it.
+        eprintln!("WARNING: v8::OwnedIsolate for snapshot was leaked");
+      } else {
+        ManuallyDrop::drop(&mut self.v8_isolate);
+      }
+    }
+  }
+}
+
 /// A single execution context of JavaScript. Corresponds roughly to the "Web
-/// Worker" concept in the DOM. A JsRuntime is a Future that can be used with
+/// Worker" concept in the DOM. A JsRuntimeImpl is a Future that can be used with
 /// an event loop (Tokio, async_std).
```

Comment:
The second sentence can be removed now - it's not longer a future, it has a `JsRuntime::run_event_loop` method